xray/client/xray_viewer.ts

```diff
@@ -0,0 +1,139 @@
+'use strict';
+
+function matrixToString(m: THREE.Matrix4): string {
+  const me = m.elements;
+  return [
+    me[0].toFixed(10),
+    me[1].toFixed(10),
+    me[2].toFixed(10),
+    me[3].toFixed(10),
+    me[4].toFixed(10),
+    me[5].toFixed(10),
+    me[6].toFixed(10),
+    me[7].toFixed(10),
+    me[8].toFixed(10),
+    me[9].toFixed(10),
+    me[10].toFixed(10),
+    me[11].toFixed(10),
+    me[12].toFixed(10),
+    me[13].toFixed(10),
+    me[14].toFixed(10),
+    me[15].toFixed(10),
+  ].join(',');
+}
+
+class NodeData {
+  public plane: THREE.Mesh;
+  public inScene: boolean;
+
+  constructor(public id: string, public boundingRect: any) {
+    this.inScene = false;
+  }
+
+  public setTexture(texture: THREE.Texture) {
+    var geometry = new THREE.PlaneGeometry(this.boundingRect['edge_length'], this.boundingRect['edge_length'], 1);
+    geometry.applyMatrix(new THREE.Matrix4().makeTranslation(this.boundingRect['edge_length'] / 2., - this.boundingRect['edge_length'] / 2., 0.));
+    var material = new THREE.MeshBasicMaterial({ map: texture, side: THREE.DoubleSide });
+    this.plane = new THREE.Mesh(geometry, material);
+    this.plane.scale.y = -1;
+    this.plane.position.set(this.boundingRect['min_x'], this.boundingRect['min_y'], 0.);
+  }
+};
+
+export class XRayViewer {
+  private nodes: { [key: string]: NodeData } = {};
+  private currentlyLoading: number;
+  private nodesToLoad: string[] = [];
+
+  constructor(private scene: THREE.Scene, private meta: any) {
+    this.currentlyLoading = 0;
+  }
+
+  public frustumChanged(matrix: THREE.Matrix4, pixelsPerMeter: number) {
+      // Figure out which view level represents our current zoom.
+      let full_size_considering_zoom = pixelsPerMeter * this.meta['bounding_rect']['edge_length'];
+      let level = 0;
+      let edge_length_px_for_level = this.meta['tile_size'];
+      while (edge_length_px_for_level < full_size_considering_zoom && level < this.meta['deepest_level']) {
+        edge_length_px_for_level *= 2;
+        level += 1;
+      }
+
+    // ThreeJS is column major.
```

Comment:
Might want to just say that matrices are encoded here in column-major order. AFAIK Three stores matrices internally in column-major, but some APIs (e.g. `set()`) use row-major.